mode/gap/gap.js

```diff
@@ -0,0 +1,207 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
+// Distributed under an MIT license: http://codemirror.net/LICENSE
+
+// based on the python codemirror mode
+(function(mod) {
+  if (typeof exports == "object" && typeof module == "object") // CommonJS
+    mod(require("../../lib/codemirror"));
+  else if (typeof define == "function" && define.amd) // AMD
+    define(["../../lib/codemirror"], mod);
+  else // Plain browser env
+    mod(CodeMirror);
+})(function(CodeMirror) {
+  "use strict";
+
+  function wordRegexp(words) {
+    return new RegExp("^((" + words.join(")|(") + "))\\b");
+  }
+
+  var wordOperators = wordRegexp(["and", "or", "not", "is", "mod"]);
+  var gapKeywords = [
+        "and", "atomic", "break", "continue", "do"
+      , "elif", "else", "end", "false", "fi", "for"
+      , "function", "if", "in", "local", "mod", "not"
+      , "od", "or", "readonly", "readwrite", "rec"
+      , "repeat", "return", "then", "true", "until"
+      , "while", "quit", "QUIT"
+      ];
+  var gapBuiltins = [ "IsBound", "Unbind", "Info", "Assert"
+                      , "TryNextMethod" ];
+
+  CodeMirror.registerHelper("hintWords", "gap", gapKeywords.concat(gapBuiltins));
+
+  function top(state) {
+    return state.scopes[state.scopes.length - 1];
+  }
+
+  CodeMirror.defineMode("gap", function(conf, parserConf) {
+    var ERRORCLASS = "error";
+
+    var singleDelimiters = parserConf.singleDelimiters || /^[\(\)\[\]\{\},:`=;\.]/;
+    var doubleOperators = parserConf.doubleOperators || /^([!<>]==|<>|<<|>>|\/\/|\*\*)/;
+    var doubleDelimiters = parserConf.doubleDelimiters || /^(\!.|\!\[)/;
+
+    var singleOperators = parserConf.singleOperators || /^[\+\-\*\/%&|\^~<>!]/;
+    var identifiers = parserConf.identifiers|| /^[_A-Za-z0-9\\*]*[_A-Za-z\\*][_A-Za-z0-9\\*]*/;
+
+    var myKeywords = gapKeywords, myBuiltins = gapBuiltins;
+    var stringPrefixes = new RegExp("^(([rub]|(ur)|(br))?('{3}|\"{3}|['\"]))", "i");
+    var keywords = wordRegexp(myKeywords);
+    var builtins = wordRegexp(myBuiltins);
+
+    if (conf) {};
+
+    // tokenizers
+    function tokenBase(stream, state) {
+      if (stream.eatSpace()) return null;
+      var ch = stream.peek();
+
+      // Handle Comments
+      if (ch == "#") {
+        stream.skipToEnd();
+        return "comment";
+      }
+       // Handle operators and Delimiters
+        if (stream.match(doubleDelimiters))
```

Comment:
Bad indentation here (and in the if statements below). (Figure out how to auto-indent in your editor, so that you don't have to think about such things.)